### PR TITLE
linalg: least squares

### PR DESCRIPTION
**Pull request by perazz**
Least squares solution to to system $Ax=b$, i.e. such that the 2-norm $||b-Ax||$ is minimized., based on LAPACK `*GELSD` functions.

- [x] base implementation
- [x] tests
- [x] exclude unsupported `xdp` 
- [x] documentation  
- [x] submodule 
- [x] examples
- [x] `subroutine` interface 

Prior art: 
- NumPy: `lstsq(a, b, rcond='warn')`
- Scipy: `lstsq(a, b, cond=None, overwrite_a=False, overwrite_b=False, check_finite=True, lapack_driver=None)`
- IMSL: `Result = IMSL_QRSOL(B, [A] [, AUXQR] [, BASIS] [, /DOUBLE] [, QR] [, PIVOT] [, RESIDUAL] [, TOLERANCE])`

Proposed implementation: generic interface for either 1 RHS or many RHS's

- `x = lstsq(a,b)` -> simplest call
- `x = lstsq(a,b,cond,overwrite_a,rank,err)` -> expert call: 
    - `real(kind(a)), intent(in), optional :: cond` = optional cutoff for rank evaluation: singular values `s(i)<=cond*maxval(s)` are considered 0. Default: `0.0`
    - `logical(lk), optional, intent(in) :: overwrite_a` = option to avoid internal allocation, but destroy input matrix `a`. Default: `.false.`
    - `integer(ilp), optional, intent(out) :: rank` = return rank of `A`, if requested
    - `type(linalg_state_type), optional, intent(out) :: err` = return state variable

cc: @jvdp1 @jalvesz @fortran-lang/stdlib I believe this is ready for consideration.

### CODE REVIEW

**Review by gnikit — APPROVED** [review]
LGTM. Maybe some comments about the chosen workspace sizes in `...gesv_space...` subroutine would help with future maintenance or performance issues.

Is there anything else that is missing, or are we happy to merge?

I would really like to follow the same pattern of issuing a Release after the merge, is there a reason why we should not do that?

**Review by jalvesz — CHANGES_REQUESTED** [review]

File: src/stdlib_linalg_least_squares.fypp
Comment (perazz) [reply src/stdlib_linalg_least_squares.fypp]: [56bf7f7](https://github.com/fortran-lang/stdlib/pull/801/commits/56bf7f7a7ba5eb7751d1d60463fb7f5ab9e50cbb)

File: src/stdlib_linalg_least_squares.fypp
Comment (perazz) [reply src/stdlib_linalg_least_squares.fypp]: [cd9b3e6](https://github.com/fortran-lang/stdlib/pull/801/commits/cd9b3e64b34a7788e917523ef475dd9b749c17ae)

File: src/stdlib_linalg_least_squares.fypp
Comment (perazz) [reply src/stdlib_linalg_least_squares.fypp:66]: The default functional style here is chosen as it is in line with the approach taken by the most used libraries. I definitely support having one more interface using a `subroutine`. In this case, what should be called like? Do you think it could be a good idea to open a separate PR with related discussion on it?

Comment (jalvesz) [reply src/stdlib_linalg_least_squares.fypp:66]: >In this case, what should be called like?

I'm not sure if there is already a convention within stdlib regarding an implementation having both a function and a subroutine interfaces simultaneously ... could be something like `slstsq` ? too many letters I feel... when working on `to_num` I used `*_base` where base where all subroutines, which are called by the function interface, so another idea:  `lstsq_base` ?

>Do you think it could be a good idea to open a separate PR with related discussion on it?
That can be a good solution to avoid holding this PR ;)

File: test/linalg/test_linalg_lstsq.fypp
Comment (jvdp1) [inline test/linalg/test_linalg_lstsq.fypp:46]: An alternative, and maybe more general, test would be to:
1. Draw a random matrix `A` and a random vector `x`
2. Compute rhs as `b= matmul(A,x)
3. Perform `p=lstsq(A,b)`
4. Check if `p == x`

**Review by jvdp1 — COMMENTED** [review]
Thank you for this PR. Here are some suggestions.

File: src/stdlib_linalg_least_squares.fypp
Comment (perazz) [reply src/stdlib_linalg_least_squares.fypp:66]: I was trying to see if I could cook a quick subroutine interface. However, we also need to decide what to do with internal allocation (can be significant): 

https://github.com/fortran-lang/stdlib/blob/90f553f0b28c0139ca069524be7cd8e37197ab6d/src/stdlib_linalg_least_squares.fypp#L156-L162

As long as this is always allocated in place, re-using `x` will not lead to significant performance benefits. So yes, I would defer this to a separate PR!

File: doc/specs/stdlib_linalg.md
Comment (perazz) [reply doc/specs/stdlib_linalg.md:623]: They can be both, but more importantly, `a` is not necessarily square

File: example/linalg/example_lstsq.f90
Comment (perazz) [reply example/linalg/example_lstsq.f90]: [09f1a5b](https://github.com/fortran-lang/stdlib/pull/801/commits/09f1a5b53a075938e699ebc54bd7d62222a1a549)

File: test/linalg/test_linalg_lstsq.fypp
Comment (perazz) [reply test/linalg/test_linalg_lstsq.fypp:46]: [63d297e](https://github.com/fortran-lang/stdlib/pull/801/commits/63d297ecdc961c44f53b451f409828d69b9f8dda)

File: doc/specs/stdlib_linalg.md
Comment (perazz) [reply doc/specs/stdlib_linalg.md]: [dafa081](https://github.com/fortran-lang/stdlib/pull/801/commits/dafa081dabbdb1143e696da030d03e5c8a0e865e)

**Comment by perazz** [comment]
Thank you all @jvdp1 @gnikit @jalvesz for the comments! 
I believe it's polished enough now. 
Regarding the allocation question @jalvesz, please let me know. Because there are 3/4 allocations per case, the subroutine interface would look similar to `gesv`: something like `call lstsq_solve(A,b,x,rwork,iwork,cwork,rank,err)` so if one wants to handle the solution fully controlling it, maybe (thinking aloud) they could just use the `gesv` LAPACK interface that we also provide? Otherwise, I'd say let's open a separate issue for this, so we can discuss further and find a suitable common ground

**Comment by jalvesz** [comment]
>Regarding the allocation question @jalvesz, please let me know. Because there are 3/4 allocations per case, the subroutine interface would look similar to gelsd: something like call lstsq_solve(A,b,x,rwork,iwork,cwork,rank,err) so if one wants to handle the solution fully controlling it, maybe (thinking aloud) they could just use the gelsd LAPACK interface that we also provide?

Indeed, full control on that regard can be done using the `gelsd` interface directly. 

>I'd say let's open a separate issue for this, so we can discuss further and find a suitable common ground

Yes, let's keep this idea in mind for latter on.

File: src/stdlib_linalg.fypp
Comment (jvdp1) [inline src/stdlib_linalg.fypp:15]:
```suggestion
  use stdlib_linalg_state, only: linalg_state_type, linalg_error_handling
```

File: src/stdlib_linalg_least_squares.fypp
Comment (jvdp1) [reply src/stdlib_linalg_least_squares.fypp:66]: >I'm not sure if there is already a convention within stdlib regarding an implementation having both a function and a subroutine interfaces simultaneously

I am not aware of such conventions. However, I am in favour of providing both functions and subroutines when it is possible, and especially when large arrays might be involved..

**Review by jvdp1 — APPROVED** [review]
Thank you @perazz . LGTM. I only have a couple of minor suggestions.

**Comment by perazz** [comment]
@jvdp1 @jalvesz @gnikit I added the subroutine interface. This interface is now fare more complicated, so I would like to ask your help deciding if this allowing the expert user has now too many control knobs? We now also have to export an interface to let the user compute what's the minimum working array space for all cases. 

Here is the new interfaces I've prepared: 

https://github.com/fortran-lang/stdlib/blob/2e370f45438a637c90b608609f6edc0ddbdc6ef6/src/stdlib_linalg.fypp#L344-L351

https://github.com/fortran-lang/stdlib/blob/2e370f45438a637c90b608609f6edc0ddbdc6ef6/src/stdlib_linalg.fypp#L295-L321

**Comment by jalvesz** [comment]
I'll try to take a close look at this soon @perazz, great work!! This makes me think about the issues when one wants to solve sparse linear systems with different preconditioners, one also needs to create a `work space` for the preconditioner. So I think it is worth defining this notion properly now.

File: src/stdlib_linalg_least_squares.fypp
Comment (perazz) [reply src/stdlib_linalg_least_squares.fypp:105]: I like your idea @jalvesz. To save us from the hassle of doing this for every single function, should we envision a storage space DT meant to work with *all* linear algebra functions? Something like: 

```
type :: linalg_storage
   integer(ilp), allocatable :: iwork(:)
   real(sp), allocatable :: rwork_sp(:) 
   real(dp), allocatable :: rwork_dp(:)
   real(qp), allocatable :: rwork_qp(:)
   complex(sp), allocatable :: cwork_sp(:) 
   complex(dp), allocatable :: cwork_dp(:)
   complex(qp), allocatable :: cwork_qp(:)
   ...
end type
```

There is no way to avoid embedding all possible datatypes, but we could at least simplify the initialization: 

```
type(linalg_storage) :: work
work = linalg_storage('solve',A,b)
work = linalg_storage('lstsq',A,b)
...
```

Comment (jalvesz) [reply src/stdlib_linalg_least_squares.fypp:105]: You could split it by kind:
```Fortran
type, abstract :: linalg_storage
   integer(ilp), allocatable :: iwork(:)
   integer(ilp), allocatable :: pivot(:)
end type

type, extends(linalg_storage) :: linalg_storage_sp
   real(sp), allocatable :: diagonals(:) 
   real(sp), allocatable :: work(:) 
end type

type, extends(linalg_storage) :: linalg_storage_dp
   real(dp), allocatable :: diagonals(:) 
   real(dp), allocatable :: work(:) 
end type

type, extends(linalg_storage) :: linalg_storage_csp
   complex(sp), allocatable :: diagonals(:) 
   complex(sp), allocatable :: work(:) 
end type

type, extends(linalg_storage) :: linalg_storage_cdp
   complex(sp), allocatable :: diagonals(:) 
   complex(sp), allocatable :: work(:) 
end type
```

And for the user he would need to select according to the working precession:
```
type(linalg_storage_sp) :: work
work = linalg_storage('solve',A,b) !> or  linalg_storage('solve',num_rows,num_cols,num_rhs) ?
work = linalg_storage('lstsq',A,b) !> or  linalg_storage('solve',num_rowsnum_cols,num_rhs) ?
```

Comment (perazz) [reply src/stdlib_linalg_least_squares.fypp:105]: Respectfully, I believe this looks a bit overengineered. I'm not sure if the user would be happy to check what version of the derived type they're using every time. Keep in mind that internally, the sizes of all necessary arrays will always be checked. So if the user tries to submit an invalid storage variable to a problem, an error will be returned. 

However, I also agree with you that having so many `allocatable` arrays inside it is also pretty dumb/non-optimal (PS: `complex` cases need 3: real, integer, complex storage). The best way to me would be to have a parameterized derived type. We may try it as we don't need type-bound procedures, but I don't know if compiler support is wide enough: 

```fortran
type :: linalg_storage(wp)
   integer, kind :: wp
   integer(ilp), allocatable :: pivot(:)
   real(wp), allocatable :: diagonals(:) 
   real(wp), allocatable :: rwork(:) 
   complex(wp), allocatable :: cwork(:) 
end type
```

So at the end of the day, maybe the big derived type is the best tradeoff?

Comment (jalvesz) [reply src/stdlib_linalg_least_squares.fypp:105]: Right, the PDT seems like the ideal solution, I'm also worried about the support in compilers, but keepings its use plain should allow for a broad use. 
>So at the end of the day, maybe the big derived type is the best tradeoff?

Probably.

**Comment by jvdp1** [comment]
Thank you @perazz . It sounds ready to be merged IMO. Do you agree? @jalvesz @gnikit ?

**Comment by gnikit** [comment]
LGTM, however some CI MacOS tests are failing. Do we know what is causing that?

**Comment by jalvesz** [comment]
Those are fails from before this merged PR https://github.com/fortran-lang/stdlib/pull/807 by @perazz, I think running the tests again would be successful.

**Comment by jvdp1** [comment]
Thank you @perazz /all. I will merge this PR now, and I suggest to wait on #806 to create a new release.

**Comment by perazz** [comment]
Great! Thank you @jvdp1 @jalvesz @gnikit. For the new release, how about waiting until most of the linear algebra functions are in? But, your proposal is also ok to me.

**Comment by perazz** [comment]
If we look at (NumPy.linalg)[https://numpy.org/doc/stable/reference/routines.linalg.html#module-numpy.linalg] for example, good LAPACK-based functions would be: 
- eigenvalues (eig, eigh, eigvals, eigvalsh)
- condition number (cond)
- matrix decompositions (svd, cholesky, qr)
- pseudo-inverse (based on svd)

**Comment by jvdp1** [comment]
We could follow the same subsections ("Matrix and vector products", "decompositions", "matrix eigenvalues", "norms", "solving equations") as in Numpy. Each section could be related with a new release. However, this would need some orders of implementation and impact your plans, which I don't want. So, as you prefer regarding releases.

**Comment by gnikit** [comment]
> For the new release, how about waiting until most of the linear algebra functions are in? But, your proposal is also ok to me.

For me the answer depends on how long it will take for all the linear algebra functions to be completed and the final release to be issued. My original thought was to aim for quick feedback loops from the users/community to inform future linalg API PRs.
The way I see it, incremental releases for high quality, self-contained work, come at a small cost to development. This is how I would classify these last few PRs.

However, the main objective is to deliver the linear algebra API not to follow release best-practices, so I leave this entirely up to you @perazz.

**Comment by perazz** [comment]
Makes a lot of sense @gnikit. If you'd like to prepare a new release already @jvdp1, I'm in for that, do let me know if/what inputs do you need from my side. Thanks!

**Pull request merged**
